_posts/2016-10-12-edsls-part-2.md

```diff
@@ -0,0 +1,413 @@
+---
+layout: post
+title: EDSLs as functions
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+*This is the second of a series of articles on "Monadic EDSLs in Scala."*
+
+Perhaps the most direct way to start writing an EDSL is to start writing
+functions. Let's say we want a language for talking about sets of integers.
+
+```scala
+trait SetLang {
+  def add(i: Int, set: Set[Int]): Set[Int]
+  def remove(i: Int, set: Set[Int]): Set[Int]
+  def exists(i: Int, set: Set[Int]): Boolean
+}
+```
+
+This works.. to the extent that we want only to work with
+`scala.collection.Set`s. As it stands we cannot talk about
+other implementations of sets, bloom filters, or sets controlled by
+remote programs. Our language isn't *abstract* enough, so let's remove
+all traces of `Set`.
+
+```scala
+trait SetLang[F[_]] {
```

Comment:
Such a type parameter would be \*-kinded in a sensible refinement of this algebra in a direction other than the one you're about to take. 😊 You're paving the way for the monadic variant, but as a foundation, this one's a little shaky.